### PR TITLE
[Improvement] Supports MR application  re-execute

### PR DESCRIPTION
**Pull request by jerqi**
### What changes were proposed in this pull request?
We increase a extra configuration to store rss configuration information, we don't modify the job.xml on the HDFS any more.

### Why are the changes needed?
If we don't have this patch, mr application cannot re-execute successfully. We will encounter the exception

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
new UT

### CODE REVIEW

File: client-mr/src/main/java/org/apache/hadoop/mapreduce/RssMRUtils.java [hunk 1]
```diff
@@ -45,22 +45,31 @@ public class RssMRUtils {
 
   // Class TaskAttemptId have two field id and mapId, rss taskAttemptID have 21 bits,
   // mapId is 19 bits, id is 2 bits.
-  public static long convertTaskAttemptIdToLong(TaskAttemptID taskAttemptID) {
+  public static long convertTaskAttemptIdToLong(TaskAttemptID taskAttemptID, int appAttemptId) {
```
Comment (colinmjj) [inline client-mr/src/main/java/org/apache/hadoop/mapreduce/RssMRUtils.java:48]: update comment to describe how to get taskId

Comment (jerqi) [reply client-mr/src/main/java/org/apache/hadoop/mapreduce/RssMRUtils.java:49]: done

File: client-mr/src/test/java/org/apache/hadoop/mapreduce/task/reduce/EvenFetcherTest.java [hunk 2]
```diff
@@ -207,20 +208,20 @@ public void obsoletedAndTipFailedEventFetch() throws IOException {
...
+      new RssEventFetcher(tid, umbilical, jobConf, MAX_EVENTS_TO_FETCH, 1);
 
     Roaring64NavigableMap expected = Roaring64NavigableMap.bitmapOf();
     for (int mapIndex = 0; mapIndex < mapTaskNum; mapIndex++) {
       if (!tipFailed.contains(mapIndex) && !obsoleted.contains(mapIndex)) {
         long rssTaskId = RssMRUtils.convertTaskAttemptIdToLong(
-          new TaskAttemptID("12345", 1, TaskType.MAP, mapIndex, 0)
-        );
+          new TaskAttemptID("12345", 1, TaskType.MAP, mapIndex, 0),
+            1);
```
Comment (frankliee) [inline client-mr/src/test/java/org/apache/hadoop/mapreduce/task/reduce/EvenFetcherTest.java:218]: unaligned

File: client-mr/src/main/java/org/apache/hadoop/mapreduce/RssMRUtils.java [hunk 3]
```diff
@@ -44,23 +44,33 @@ public class RssMRUtils {
...
-                                                    long rssTaskAttemptId) {
+  public static TaskAttemptID createMRTaskAttemptId(
+      JobID jobID,
+      TaskType taskType,
+      long rssTaskAttemptId,
+      int appAttemptId) {
+    if (appAttemptId < 1) {
+      throw new RssException("appAttemptId " + appAttemptId + " is wrong");
+    }
     TaskID taskID = new TaskID(jobID, taskType, (int)(rssTaskAttemptId & MAX_TASK_ID));
```
Comment (frankliee) [inline client-mr/src/main/java/org/apache/hadoop/mapreduce/RssMRUtils.java:72]: 1024  or << 10

File: client-mr/src/main/java/org/apache/hadoop/mapreduce/task/reduce/RssEventFetcher.java [hunk 4]
```diff
@@ -47,16 +47,19 @@
...
   private int tipFailedCount = 0;
   private final int totalMapsCount;
+  private final int appAttemptId;
 
   public RssEventFetcher(TaskAttemptID reduce,
                          TaskUmbilicalProtocol umbilical,
                          JobConf jobConf,
-                         int maxEventsToFetch) {
+                         int maxEventsToFetch,
+                         int appAttemptId) {
```
Comment (frankliee) [inline client-mr/src/main/java/org/apache/hadoop/mapreduce/task/reduce/RssEventFetcher.java:56]: Put it to the first argument position

File: client-mr/src/main/java/org/apache/hadoop/mapreduce/RssMRUtils.java (hunk 3 above)
Comment (jerqi) [reply client-mr/src/main/java/org/apache/hadoop/mapreduce/RssMRUtils.java:72]: MapReduce use 1000.

Comment (frankliee) [reply client-mr/src/main/java/org/apache/hadoop/mapreduce/RssMRUtils.java:72]: ok

File: client-mr/src/main/java/org/apache/hadoop/mapreduce/task/reduce/RssEventFetcher.java (hunk 4 above)
Comment (jerqi) [reply client-mr/src/main/java/org/apache/hadoop/mapreduce/task/reduce/RssEventFetcher.java:56]: What‘s your reason?

File: client-mr/src/main/java/org/apache/hadoop/mapreduce/RssMRConfig.java [hunk 5]
```diff
@@ -115,6 +115,8 @@ public class RssMRConfig {
   public static String RSS_ACCESS_TIMEOUT_MS = MR_RSS_CONFIG_PREFIX + RssClientConfig.RSS_ACCESS_TIMEOUT_MS;
   public static int RSS_ACCESS_TIMEOUT_MS_DEFAULT_VALUE = RssClientConfig.RSS_ACCESS_TIMEOUT_MS_DEFAULT_VALUE;
 
+  public static String RSS_EXTRA_CONF_FILE = "extra.xml";
```
[the remark this answers is not in the document]
Comment (jerqi) [reply client-mr/src/main/java/org/apache/hadoop/mapreduce/RssMRConfig.java:118]: ok, more reasonable name.

File: client-mr/src/main/java/org/apache/hadoop/mapreduce/task/reduce/RssEventFetcher.java (hunk 4 above)
Comment (jerqi) [reply client-mr/src/main/java/org/apache/hadoop/mapreduce/task/reduce/RssEventFetcher.java:56]: OK

**Comment by jerqi** [comment]
@frankliee @colinmjj All comments are addressed.

**Comment by frankliee** [comment]
LGTM

**Review by frankliee — APPROVED** [review]

**Pull request merged**